### ISSUE TITLE
Include :: Multi-level include doesn't work with collections of type List<T> etc (incorrect overload of Include chosen by compiler)

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFramework#1463]
Repro:

```

    public class Customer
    {
        public int Id { get; set; }
        public List<Order> Orders { get; set; }
    }

    public class Order
    {
        public int Id { get; set; }
        public List<OrderLine> OrderLines { get; set; }
    }

    public class OrderLine
    {
        public int Id { get; set; }
    }

    public class MyContext : DbContext
    {
        public DbSet<Customer> Customers { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Customer>().HasMany(c => c.Orders);
            modelBuilder.Entity<Order>().HasMany(o => o.OrderLines);
        }

        protected override void OnConfiguring(DbContextOptions options)
        {
            options.UseSqlServer(new SqlConnectionStringBuilder { DataSource = ".", InitialCatalog = "Repro1463", MultipleActiveResultSets = true, IntegratedSecurity = true }.ToString());
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureCreated();
                
                // won't compile
                ctx.Customers.Include(c => c.Orders).ThenInclude(o => o.OrderLines);
            }
        }
    }

```

**Comment by bricelam** [comment]
Weird. I wonder if it's a Roslyn bug. I was seeing similar problems with type parameter inference a couple of weeks ago on the relationship fluent APIs.

**Comment by bricelam** [comment]
Oh, maybe it has something to do with portable too...

**Comment by anpete** [comment]
Looks like its just standard overload resolution. I think we have at least two options:

1) Rename the collection version. E.g. "ThenIncludeCollection", or perhaps: "With"/"WithCollection"; or even "Expand".
2) Add further overloads for common collection types. Not ideal...

cc @divega

**Comment by maumar** [comment]
You would also need to rename Include as well, not only ThenInclude. Still better than option 2.

**Comment by anpete** [comment]
Actually, IncludeOne/Many would be consistent with CF.

**Comment by divega** [comment]
@anpete @maumar FYI, I was able to make an older version of the C# compiler match this particular case by having a version of the method with this signature:

``` C#
public static IIncludableQueryable<TEntity, TProperty> ThenInclude<TEntity, TPreviousProperty, TProperty>(
    this IIncludableQueryable<TEntity, ICollection<TPreviousProperty>> source,
    Expression<Func<TPreviousProperty, TProperty>> collectionNavigationPropertyPath)
    where TEntity : class
{
```
Notice the type of the source argument will match if the previous call to `Include` landed on a collection navigation, which is actually the case in which the `ThenInclude` method is required, as far as I know. 

For this to work I also had to change `IIncludableQueryabe` to be covariant instead of contravariant with respect to `TProperty` e.g.:

``` C#
public interface IIncludableQueryable<out TEntity, out TProperty> : IQueryable<TEntity>
{

}
```

I tried to compile this overload alongside the existing ones with several queries against a modified version of this model (I added back reference nav props) and there was no ambiguity. That said, I haven't tried to compile or run any tests with this change or confirm that the variance change had no impact (still trying to get to a good build of VS 2015 working on my machine).
